### PR TITLE
Fix encryption hierarchy issues with zfs recv -d

### PR DESCRIPTION
**Pull request by tcaputi**
Currently, the recv_fix_encryption_hierarchy() function accepts
'destsnap' as one of its parameters. Oriinally, this was intended
to be the top-level dataset of a receive (whether or not the
receive was recursive). Unfortunately, this parameter actually is
simply the input that is passed in from the command line. When
the user specifies 'zfs recv -d', this string is actually only the
name of the receiving pool since the rest of the name is derived
from the send stream. This causes the function to fail, leaving
some datasets with an invalid encryption hierarchy.

This patch resolves this problem by passing in the top_zfs variable
instead. In order to make this work, this patch also includes some
changes that ensure the value is always present when we need it.

Fixes: #9273

Signed-off-by: Tom Caputi <tcaputi@datto.com>

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by tcaputi** [comment]
@behlendorf I think all of the remaining test failures here are spurious ones from upstream.

**Comment by tcaputi** [comment]
Yeah, I probably should add a test case to this. I'll do that.

File: lib/libzfs/libzfs_sendrecv.c [hunk 1]
```diff
@@ -867,7 +867,7 @@ send_iterate_fs(zfs_handle_t *zhp, void *arg)
 
 	tosnap_txg = get_snap_txg(zhp->zfs_hdl, zhp->zfs_name, sd->tosnap);
 	if (tosnap_txg != 0)
-		sd->tosnap_txg = tosnap_txg;
+		sd->tosnap_txg = tosnap_txg+1;
```
Comment (loli10K) [inline lib/libzfs/libzfs_sendrecv.c]: Is this change related in any way to the original issue?

File: lib/libzfs/libzfs_sendrecv.c [hunk 2]
```diff
@@ -4862,6 +4864,9 @@ zfs_receive_one(libzfs_handle_t *hdl, int infd, const char *tosnap,
 	    &oxprops, &wkeydata, &wkeylen, errbuf)) != 0)
 		goto out;
 
+	if ((newfs || stream_avl) && type == ZFS_TYPE_FILESYSTEM && !redacted)
```
Comment (loli10K) [inline lib/libzfs/libzfs_sendrecv.c]: Can we move this after the comment at line 5099?

> Mount the target filesystem (if created).  Also mount any
> children of the target filesystem if we did a replication
> receive (indicated by stream_avl being non-NULL).

File: lib/libzfs/libzfs_sendrecv.c (hunk 1 above)
Comment (tcaputi) [reply lib/libzfs/libzfs_sendrecv.c]: Sorry. My coworker thought he would be funny and added that when I walked away for a second. Thanks for catching it.

**Comment by tcaputi** [comment]
@loli10K your comments have been addressed.

File: tests/zfs-tests/tests/functional/cli_root/zfs_receive/zfs_receive_raw_-d.ksh [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+#
+# Copyright (c) 2019 Datto, Inc. All rights reserved.
+#
+
+. $STF_SUITE/include/libtest.shlib
+
+#
+# DESCRIPTION:
+# zfs receive -f should create the expected encryption hierarchy.
```
Comment (loli10K) [inline tests/zfs-tests/tests/functional/cli_root/zfs_receive/zfs_receive_raw_-d.ksh]: nit: sorry, i somehow missed this in my first review: i think this is supposed to be _zfs receive -d_

Comment (tcaputi) [reply tests/zfs-tests/tests/functional/cli_root/zfs_receive/zfs_receive_raw_-d.ksh]: Thats fine. I missed it when writing it :) fixed.
